### ISSUE TITLE
Adding an image to a header or footer

### ISSUE THREAD

**Issue by franktudor** [cfsimplicity/lucee-spreadsheet#101]
For my client we want to add a small logo to the left header.  Right now we can add text which is great and with the `&Helpers` (that i just discovered) we can style, color, add page numbers, etc. But one thing I was unable to do was to place an image in the header. This could be because I don't understand how.  The other minor issue is that I cannot achieve a line break `\n`.

I did find another third party tool that gave some really good explanations on those `&Helpers`

[http://xlsxwriter.readthedocs.io/example_headers_footers.html ](http://xlsxwriter.readthedocs.io/example_headers_footers.html )

**So using what you have available plus the &Helpers (in the link above) allow me to do things like this:**

`spreadsheet.setHeader(workbook=workbook,`
 		                   ` leftHeader="Subset Testing Inc.",`
 		                    `centerHeader='&"Arial,Bold"Page &P of &N',`
 		                    `rightHeader="Methods and Assumptions" );`
`spreadsheet.setFooter(workbook=workbook,`
 		                    `leftFooter="Version 1.0",`
 		                    `centerFooter='&"Arial,Bold"&KFF0000Confidential Competitive Information',`
 		                    `rightFooter="Critical Anomalies" );`

- **&"Arial,Bold"** (Picks a font type and Sets to Bold Style)
- **&P** (Page)
- **&N** (Total pages. Useful with &P to give example page 1 of (total pages))
- **&K** (&KFF0000 makes my font red)

But I cannot attach an image (that Subset testing Inc. is where I need that .png graphic). Maybe you can help here? It'll be my last request (I feel like I am asking a lot).  Maybe I could help you with documentation and provide advanced examples to return the favor? I have a pretty robust build currently working in XLSX.

**Comment by cfsimplicity** [comment]
I wasn't aware of these formatting codes. Very useful. But from what I've read it doesn't seem possible to insert images. There is a code `&G` for "picture as a background", but I can't see how to supply the image itself to fill the placeholder. The `set_header` method in the Python library you mention has a second argument for the image details, but the POI function only allows a single string argument according to the docs. Googling only seems to turn up others equally puzzled.

I'll leave the issue open for a while though in case a solution comes up.

To get new lines it seems you can just use CFML:
```
spreadsheet.setHeader(
 workbook=workbook
 ,leftHeader="First line#Chr(13)#Second line"
);
```
Please don't stop making requests. My own use of the library is mainly just for basic data dumps so I rely on others to suggest more advanced features that they genuinely have a need for. I can't promise to fulfill them all, but I'm always happy to improve the library if I can.

**Comment by franktudor** [comment]
Your library is excellent! Clean too. Thanks for the tip on the line break. That image thing is nagging me because they really want it, but they will accept a solution without it so we can let it simmer, maybe something will reveal itself over time. Keep up the great work!

**Comment by GaryStanton** [comment]
Very late to the party, but I was looking at trying to do this recently.
In the end, I simply created a template xlsx with the images in there, loaded them up with LuceeSpreadsheet and added the rows from a query manually. That works well...
However, if you're looking to create an image dynamically - which I may have to in the not too distant future - it _should_ be possible, since an xlsx file is an archive and contains the image and pointers to it in XML. It would theoretically be a case of unzipping the file, messing with the contents and rezipping. All should be possible with native CFML functionality.

I found some info on this on StackOverflow, which also has an example implementation using POI:
https://stackoverflow.com/questions/51077404/apache-poi-adding-watermark-in-excel-workbook/51103756#51103756

**Comment by cfsimplicity** [comment]
Really useful info, Gary, thanks. I'll talk a look at that SO code when I get a chance.

**Comment by cfsimplicity** [comment]
Well, that was quite a challenge, but I think I've got it working with help from that StackOverflow answer.

@GaryStanton, @franktudor could you please have a go with the `develop` branch which now has these two new methods?
```
setHeaderImage( workbook, position, image [, imageType] );
setFooterImage( workbook, position, image [, imageType] );
```
`position` should be `left|center|right`, and `image` can be a file path or an image object. If you specify an image object you must also provide the type, e.g. `png`.

There are a couple of limitations to be aware of:

1. They will only work with XLSX files.
2. You can't add footers to an existing spreadsheet file read from disk which already has one or more header or footer images. I've tried but can't seem to get round this. Hopefully it won't be a common use case.

Also note that the lib/osgi bundle has changed to support this.

All feedback appreciated.

**Comment by cfsimplicity** [comment]
Released in v2.18.0
